File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Listeners/onCommand.java
package jp.aoichaan0513.A_TosoGame_Live.Listeners;

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI;
import jp.aoichaan0513.A_TosoGame_Live.Commands.Command.Script;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.BlockCommandSender;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.server.ServerCommandEvent;

import java.util.ArrayList;
import java.util.List;

public class onCommand implements Listener {

    private static List<String> list = new ArrayList<>();

    @EventHandler
    public void onPlayerCommand(PlayerCommandPreprocessEvent e) {
        Player p = e.getPlayer();
        String cmdName = e.getMessage().split(" ")[0].split(":").length > 1 ? e.getMessage().split(" ")[0].split(":")[1] : e.getMessage().split(" ")[0].substring(1);
        String[] args = e.getMessage().split(" ").length > 1 ? e.getMessage().substring(cmdName.length() + 1).split(" ") : new String[0];

        if (!TosoGameAPI.isAdmin(p))
            p.sendMessage(ChatColor.GRAY + "> " + e.getPlayer().getName() + ": " + e.getMessage());

        for (Player player : Bukkit.getOnlinePlayers())
            if (TosoGameAPI.isAdmin(player))
                player.sendMessage(ChatColor.GRAY + "> " + e.getPlayer().getName() + ": " + e.getMessage());

        if (list.contains("/" + cmdName)) {
            if (!TosoGameAPI.isAdmin(p)) {
                e.setCancelled(true);
                p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "実行できません。");
                return;
            }
        } else {
            if (Script.hasFile(cmdName)) {
                e.setCancelled(true);

                WorldConfig worldConfig = Main.getWorldConfig();
                if (worldConfig.getGameConfig().getScript()) {
                    p.chat("/script " + cmdName.substring(1) + e.getMessage().substring(cmdName.length()));
                    return;
                } else {
                    p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "管理者によってスクリプト機能が無効にされています。");
                    return;
                }
            }
        }
    }

    @EventHandler
    public void onServerCommand(ServerCommandEvent e) {
        if (e.getSender() instanceof BlockCommandSender) {
            /*
            Block b = ((BlockCommandSender) e.getSender()).getBlock();

            Bukkit.getConsoleSender().sendMessage(ChatColor.GRAY + "> Command Block: /" + e.getCommand() + "\n" +
                    ChatColor.GRAY + "X: " + b.getX() + ", Y: " + b.getY() + ", Z:" + b.getZ());
            */
        } else if (e.getSender() instanceof ConsoleCommandSender) {
            Bukkit.getConsoleSender().sendMessage(ChatColor.GRAY + "> Console: /" + e.getCommand());
        }
    }

    public static void addBlockCommand(String cmdName) {
        if (!list.contains("/" + cmdName))
            list.add("/" + cmdName);
    }

    public static void removeBlockCommand(String cmdName) {
        if (list.contains("/" + cmdName))
            list.remove("/" + cmdName);
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Listeners/Minecraft/onCommand.kt
package jp.aoichaan0513.A_TosoGame_Live.Listeners.Minecraft

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI
import jp.aoichaan0513.A_TosoGame_Live.Commands.Command.Script
import jp.aoichaan0513.A_TosoGame_Live.Main
import net.md_5.bungee.api.chat.ClickEvent
import net.md_5.bungee.api.chat.ComponentBuilder
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.BlockCommandSender
import org.bukkit.command.ConsoleCommandSender
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerCommandPreprocessEvent
import org.bukkit.event.server.ServerCommandEvent

class onCommand : Listener {

    @EventHandler
    fun onPlayerCommand(e: PlayerCommandPreprocessEvent) {
        val p = e.player
        val msg = e.message

        val cmdName = if (msg.split(" ")[0].split(":").size > 1) msg.split(" ")[0].split(":")[1] else msg.split(" ")[0].substring(1)

        val textComponent1 = TextComponent("${ChatColor.GRAY}> ${p.name}: ")
        val textComponent2 = TextComponent(msg)
        textComponent2.color = net.md_5.bungee.api.ChatColor.GRAY
        textComponent2.clickEvent = ClickEvent(ClickEvent.Action.SUGGEST_COMMAND, e.message)
        textComponent2.hoverEvent = HoverEvent(HoverEvent.Action.SHOW_TEXT, ComponentBuilder("${ChatColor.GRAY}${ChatColor.BOLD}クリックでコマンドをチャット欄に入力します。").create())
        textComponent1.addExtra(textComponent2)

        Bukkit.getOnlinePlayers().filter { TosoGameAPI.isAdmin(it) }.forEach {
            it.spigot().sendMessage(textComponent1)
        }

        if (set.contains("/$cmdName")) {
            if (TosoGameAPI.isAdmin(p)) return
            e.isCancelled = true
            p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}実行できません。")
        } else {
            if (Script.hasFile(cmdName)) {
                e.isCancelled = true
                val worldConfig = Main.worldConfig
                if (worldConfig.gameConfig.script) {
                    p.performCommand("script ${cmdName.substring(1)}${e.message.substring(cmdName.length)}")
                } else {
                    p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}管理者によってスクリプト機能が無効にされています。")
                }
            }
        }
    }

    @EventHandler
    fun onServerCommand(e: ServerCommandEvent) {
        if (e.sender is BlockCommandSender) {
            /*
            Block b = ((BlockCommandSender) e.getSender()).getBlock();

            Bukkit.getConsoleSender().sendMessage(ChatColor.GRAY + "> Command Block: /" + e.getCommand() + "\n" +
                    ChatColor.GRAY + "X: " + b.getX() + ", Y: " + b.getY() + ", Z:" + b.getZ());
            */
        } else if (e.sender is ConsoleCommandSender) {
            Bukkit.getConsoleSender().sendMessage("${ChatColor.GRAY}> Console: /${e.command}")
        }
    }

    companion object {
        private val set = mutableSetOf<String>()

        fun addBlockCommand(cmdName: String) {
            set.add("/$cmdName")
        }

        fun removeBlockCommand(cmdName: String) {
            set.remove("/$cmdName")
        }
    }
}